src/GlobalPayments.Api/Terminals/INGENICO/IngenicoController.cs

```diff
@@ -0,0 +1,288 @@
+            }
+        }
+
+        #region overrides
+        internal override ITerminalResponse ManageTransaction(TerminalManageBuilder builder) {
+            var request = BuildManageTransaction(builder);
+
+            if (builder.TransactionType == TransactionType.Cancel) {
+                return DoCancelRequest(request);
+            }
+            else if (builder.TransactionType == TransactionType.Reversal) {
+                return DoReverseRequest(request);
+            }
+            else {
+                return DoRequest(request);
+            }
+        }
+
+        internal override ITerminalReport ProcessReport(TerminalReportBuilder builder) {
+            var request = TerminalUtilities.BuildIngenicoRequest(INGENICO_REQ_CMD.RECEIPT.FormatWith(builder.ReceiptType), _settings.ConnectionMode);
+            return ReportRequest(request);
+        }
+
+        internal override ITerminalResponse ProcessTransaction(TerminalAuthBuilder builder) {
+            var request = BuildRequestMessage(builder);
+            return DoRequest(request);
+        }
+
+        internal override byte[] SerializeRequest(TerminalAuthBuilder builder) {
+            return BuildRequestMessage(builder).GetSendBuffer();
+        }
+
+        internal override byte[] SerializeRequest(TerminalManageBuilder builder) {
+            throw new NotImplementedException();
+        }
+
+        internal override byte[] SerializeRequest(TerminalReportBuilder builder) {
+            throw new NotImplementedException();
+        }
+        #endregion
+
+        #region Methods
+        internal IDeviceMessage BuildManageTransaction(TerminalManageBuilder builder) {
+            var _refNumber = builder.ReferenceNumber;
+            var _amount = ValidateAmount(builder.Amount);
+            var _returnRep = 1;
+            var _paymentMode = ValidatePaymentMode(builder.PaymentMode);
+            var _paymentType = (int?)((IngenicoInterface)_device).paymentMethod ?? 0;
+            var _currencyCode = (string.IsNullOrEmpty(builder.CurrencyCode) ? "826" : builder.CurrencyCode);
+            var _privData = "EXT0100000";
+            var _immediateAns = 0;
+            var _forceOnline = 0;
+            var _extendedData = "0000000000";
+
+            // For Auth code value in extended data
+            if (!string.IsNullOrEmpty(builder.AuthCode))
+                _extendedData = ValidateExtendedData(builder.AuthCode, builder.ExtendedDataTags);
+            else if (!builder.TransactionId.IsNull() && builder.TransactionType == TransactionType.Reversal)
+                // For Reversal with Transaction Id value in extended data
+                _extendedData = ValidateExtendedData(builder.TransactionId, ExtendedDataTags.TXN_COMMANDS_PARAMS);
+            else
+                _extendedData = ValidateExtendedData(builder.TransactionType.ToString(), ExtendedDataTags.TXN_COMMANDS);
+
+            string message = string
```

Comment:
The indentation is not right here. I think I would also prefer the use of a string builder over this mass string.format method.